src/Serialization.SystemTextJson/GraphQLRequestDeserializer.cs

```diff
@@ -0,0 +1,61 @@
+using GraphQL.Server.Transports.AspNetCore.Common;
+using System;
+using System.IO;
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace GraphQL.Server.Serialization.SystemTextJson
+{
+    public class GraphQLRequestDeserializer : IGraphQLRequestDeserializer
+    {
+        private readonly JsonSerializerOptions _serializerOptions = new JsonSerializerOptions();
+
+        public GraphQLRequestDeserializer(Action<JsonSerializerOptions> configure)
+        {
+            configure(_serializerOptions);
+        }
+
+        public IGraphQLRequest Default() => new GraphQLRequest();
+
+        public async Task<GraphQLRequestDeserializationResult> FromBodyAsync(Stream stream)
+        {
+            using var ms = new MemoryStream();
+            await stream.CopyToAsync(ms).ConfigureAwait(false);
+            var jsonBytes = ms.ToArray();
+
+            JsonTokenType tokenType;
+            try
+            {
+                tokenType = GetTokenType(jsonBytes);
+            }
+            catch (Exception)
+            {
+                tokenType = JsonTokenType.None;
+            }
+
+            var result = new GraphQLRequestDeserializationResult() { WasSuccessful = true };
+
+            switch (tokenType)
+            {
+                case JsonTokenType.StartObject:
+                    result.Single = JsonSerializer.Deserialize<GraphQLRequest>(jsonBytes.AsSpan(), _serializerOptions);
+                    break;
+                case JsonTokenType.StartArray:
+                    result.Batch = JsonSerializer.Deserialize<GraphQLRequest[]>(jsonBytes.AsSpan(), _serializerOptions);
+                    break;
+                default:
+                    result.WasSuccessful = false;
+                    break;
+            }
+
+            return result;
+        }
+
+        private static JsonTokenType GetTokenType(byte[] bytes)
+        {
+            var reader = new Utf8JsonReader(bytes.AsSpan());
+            var success = reader.Read();
+            return success ? reader.TokenType : JsonTokenType.None;
```

Comment:
From @sungam3r on previous code location:

> This line can outweigh all the optimizations that occur around. A potentially large object is allocated each time on the heap. I understand that it was necessary to find some way to write completely asynchronous code. Let's try to at least mitigate the problem of allocations of arrays using the array pool. The required array size can be obtained from Content-Length header. I think that using the pool here would be appropriate.

Reply:
Hey just wanted to confirm what you meant by array size. Do you mean up where it pulls `jsonBytes`? That the length of the array should be initialized with Content-Length value?

Reply:
I mean the size of rented pooled buffer.

Reply:
Sorry, really not my area of expertise. Mind making an edit yourself or giving me a little snippet?